### ISSUE TITLE
v158: Fatal error during checkout when paying fully with Gift Certificate

### ISSUE THREAD

**Issue by lat9** [zencart/zencart#4952]
Load up a customer account with sufficient Gift Certificates/Vouchers to fully pay for an order.

Place one or more items into the cart and go to checkout.  On the `checkout_payment` page, enter the full order value in the "Use Gift Certificate" value input (or whatever it's called).

Upon pressing the continue button, you're met with a blank page, with resultant logs similar to the following
```
[30-Jun-2022 14:27:54 America/New_York] Request URI: /zc158/index.php?main_page=checkout_confirmation, IP address: 127.0.0.1
#0 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php(96): zen_debug_error_handler()
#1 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
--> PHP Warning: Undefined variable $payment_modules in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php on line 96.

[30-Jun-2022 14:27:54 America/New_York] Request URI: /zc158/index.php?main_page=checkout_confirmation, IP address: 127.0.0.1
#0 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php(96): zen_debug_error_handler()
#1 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
--> PHP Warning: Attempt to read property "modules" on null in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php on line 96.

[30-Jun-2022 14:27:54 America/New_York] Request URI: /zc158/index.php?main_page=checkout_confirmation, IP address: 127.0.0.1
#0 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php(157): zen_debug_error_handler()
#1 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
--> PHP Warning: Undefined variable $ in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php on line 157.

[30-Jun-2022 14:27:54 America/New_York] PHP Fatal error:  Uncaught TypeError: method_exists(): Argument #1 ($object_or_class) must be of type object|string, null given in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php:157
Stack trace:
#0 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php(157): method_exists(NULL, 'alterShippingEd...')
#1 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
#2 {main}
  thrown in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php on line 157

[30-Jun-2022 14:27:54 America/New_York] Request URI: /zc158/index.php?main_page=checkout_confirmation, IP address: 127.0.0.1
--> PHP Fatal error: Uncaught TypeError: method_exists(): Argument #1 ($object_or_class) must be of type object|string, null given in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php:157
Stack trace:
#0 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php(157): method_exists(NULL, 'alterShippingEd...')
#1 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
#2 {main}
  thrown in C:\xampp\htdocs\zc158\includes\modules\pages\checkout_confirmation\header_php.php on line 157.
```

**Comment by lat9** [comment]
Noting that a similar issue will be raised if the order's paid with a 100%-off coupon.

**Comment by lat9** [comment]
This gets uglier and uglier.  Once the `checkout_confirmation` issues are resolved, there are now warnings during `checkout_process`:
```
[01-Jul-2022 10:35:14 America/New_York] Request URI: /zc158/index.php?main_page=checkout_process, IP address: 127.0.0.1
#0 C:\xampp\htdocs\zc158\includes\classes\payment.php(248): zen_debug_error_handler()
#1 C:\xampp\htdocs\zc158\includes\modules\checkout_process.php(91): payment->before_process()
#2 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_process\header_php.php(14): require('C:\\xampp\\htdocs...')
#3 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
--> PHP Warning: Undefined global variable $ in C:\xampp\htdocs\zc158\includes\classes\payment.php on line 248.

[01-Jul-2022 10:35:14 America/New_York] Request URI: /zc158/index.php?main_page=checkout_process, IP address: 127.0.0.1
#0 C:\xampp\htdocs\zc158\includes\classes\payment.php(266): zen_debug_error_handler()
#1 C:\xampp\htdocs\zc158\includes\modules\checkout_process.php(96): payment->after_order_create()
#2 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_process\header_php.php(14): require('C:\\xampp\\htdocs...')
#3 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
--> PHP Warning: Undefined global variable $ in C:\xampp\htdocs\zc158\includes\classes\payment.php on line 266.

[01-Jul-2022 10:35:17 America/New_York] Request URI: /zc158/index.php?main_page=checkout_process, IP address: 127.0.0.1
#0 C:\xampp\htdocs\zc158\includes\classes\payment.php(257): zen_debug_error_handler()
#1 C:\xampp\htdocs\zc158\includes\modules\pages\checkout_process\header_php.php(17): payment->after_process()
#2 C:\xampp\htdocs\zc158\index.php(35): require('C:\\xampp\\htdocs...')
--> PHP Warning: Undefined global variable $ in C:\xampp\htdocs\zc158\includes\classes\payment.php on line 257.
```

This a another chicken-and-egg case, where the `payment` class depends on the `order_total` class' setting of `$credit_covers` (in the pre_confirmation_check method) to determine whether/not a 'real' payment method is required, but the checkout-process doesn't "want" to run that check if the current payment method is a `paypal` variant.

The saga continues ...

**Comment by lat9** [comment]
See comments in (now-merged) PR #4957; re-opening.
